### DISCUSSION TITLE
Question about unlabelled dataset.

### DISCUSSION THREAD

**Discussion by tcapelle** [KeremTurgutlu/self_supervised#24] [Q&A] [answered]
Hey nice work you have put here!
As I see on the notebooks, you always train with a labelled dataset. (`tfms = [ParentLabeller, Categorize]`).
Is this really needed?
Would you mind sharing a notebooks with the pretraining on an unlabelled dataset?

**Comment by KeremTurgutlu** [comment, marked as answer]
Thank you! No it can be a dataset without labels. So you can just replace label transforms with something like `tfms=[lambda x: return 0]`. I can add a notebook later too.

Loss computation is handled inside the callback it is unsupervised.

**Comment by tcapelle** [comment]
Hey, have you tried the ladderboard without the full dataset on `fine_tune`?
I am trying the unsupervised learning on a a 500k image dataset with only 2000 annotated ones and I am having a hard time beating the supervised approach.

**Reply by KeremTurgutlu** [reply]
How many images are there in supervised version? It's expected that supervised to perform better if your full dataset is labelled. Self supervision is helpful when you have lots of unlabelled images.

You should always use **all** the annotated data you in practice, worst case that can happen is your self-supervised encoder can perform no better than a randomly initialized one, so it would be act like random initialization. Although, this is unlikely it can happen.

Sometimes imagenet pretrained or other supervised pretrained encoders can perform better if you don't have enough unlabelled data to learn good representations with unsupervised learning. But again, in your case 500k images seems enough to learn from.

**Reply by KeremTurgutlu** [reply]
If you can provide more details about the problem and dataset, maybe we can figure out why results are not good.

**Reply by tcapelle** [reply]
About the Imagewang ladderboard, what I mean, is that you used the same dataset twice.
- The full dataset for pre-training the encoder as a self-supervised task
- Then fine tune on the whole dataset but with labels.

What i would like to see is to pretrain on the whole dataset and fine tune on a small sample of annotated images and achieve high accuracy anyway. Maybe I am missing something, but the papers report this type of approach, achieving almost SOTA levels with only as little as 1% of labelled data.
For my task, I have 500k images without label and around 2000 images with label. What I see is that the pretraining on the full dataset has little impact, as training the model from scratch on the 2000 images leads to almost the same accuracy.

**Reply by KeremTurgutlu** [reply]
> About the Imagewang ladderboard, what I mean, is that you used the same dataset twice.
> 
> * The full dataset for pre-training the encoder as a self-supervised task
> * Then fine tune on the whole dataset but with labels.

This is not correct. In Imagewang leaderboard not all images are labeled there are different datasets `train` `val` and `unsup`. We use all data to do pretraining and then do downstream training on just `train`. Please refer to description of the leaderboard, and you can also take a look at the notebooks linked [here](https://github.com/fastai/imagenette#image%E7%BD%91-leaderboard).

>For my task, I have 500k images without label and around 2000 images with label. What I see is that the pretraining on the full dataset has little impact, as training the model from scratch on the 2000 images leads to almost the same accuracy.

There might be many reasons for failure in your case. You might need to dig deeper to understand.

- Your pretraining might not be learning anything.
- Your unlabelled and labelled datasets might have different distributions.
- Your code might have bug.
- More...

The whole point of Imagewang leaderboard is to leverage unlabelled data, and in my notebooks I show it otherwise they wouldn't be accepted in the leaderboard :) 

Also, other people reported successful results using this repo (better and faster results compared to official implementations), I suggest checking out [here](https://github.com/KeremTurgutlu/self_supervised/discussions/20).

Best would be to share your code/notebook so that we understand what's going on there. Training these algorithms might require hyperparameter tuning for different problem and datasets.

**Reply by KeremTurgutlu** [reply]
Also, lately I've been experimenting with medical data, chest X-ray in this case. What I did to experiment is very similar to your case. I took NIH dataset which has 80k+ images. I hold out 10k data as labelled and used remaining 70k as unlabelled. I used SWAV algorithm and pretrained an encoder with unlabelled 70k. Then I compared two things:

- 7k training with imagenet init and validation on 3k.
- 7k training with SWAV encoder init and validation on 3k.

Even after 1-2 epochs self-supervised pretrained model performed **much** better. Here is the [notebook](https://github.com/KeremTurgutlu/medical/blob/nih-cxr/nbs/04%20-%20nih_cxr.ipynb) for reference.

Unfortunately, there is no 1 rule for all when training these algorithms and I admit sometimes is not easy but you can assume it works for most cases, or should try to make it work by figuring out what might be holding it back from learning.

Hope these points help, in future if I have time I would like to help in more detail about your own specific case if you can provide the code, experiments etc..

**Comment by tcapelle** [comment]
Thanks for all this info.
My bad not getting the notebooks, indeed the get_image_files is all over
the three folders on pretext.
Indeed my unlabeled data is not as good as the labeled, and probably that's
why I see little gain.
I also have to say that my models are already very accurate, 93% so is hard
to assess gains.
Thanks for the awesome repo.
I cannot share most of my work as it is on my org private repository using
customers data.
